WordPress/gutenberg
# Refactor classic menu conversion process

## Description
Alternative approach to https://github.com/WordPress/gutenberg/pull/38824. Continues working towards addressing https://github.com/WordPress/gutenberg/issues/37190.

This PR seeks both to

- simplify the logic around converting classic menus to block based menus in the Nav block.
- avoid the impression of the UI "locking up" when the classic menu is selected and instead provide feedback to the user that an operation is taking place.

The current hooks are complex to reason about and don't expose enough information about the progress of the conversion to allow for the UI to respond (e.g. loading states...etc).

This PR builds on the approach outlined by @talldan in https://github.com/WordPress/gutenberg/pull/38824#pullrequestreview-883918504. Essentially we create a new hook which exposes:

1. A function which trigger the conversion process.
2. The state of the conversion process.

~🙏 **Please note**: This PR _does not_ attempt to solve the loading state indicator problems that are noticeable in this PR. This is being addressed in https://github.com/WordPress/gutenberg/pull/38907.~

## Outstanding problems to solve

- [x] Allow the dropdown to close immediately upon click to avoid the UI feeling like it's locking up.
- [x] ~Create loading UI states _in the block_ whilst the conversion process happens (good things we expose the state!).~ To be addressed following merge of https://github.com/WordPress/gutenberg/pull/38907.
- [x] Avoid component unmount errors when closing the dropdown whilst the async conversion process is still running.
- [x] Migrate towards usage of registry as suggest by @adamziel in https://github.com/WordPress/gutenberg/pull/38824#issuecomment-1041549542.
- [x] Handle errors from the async conversion process (`.catch()` .etc).
- [x] Provide accessible progress announcements of conversion status.

I will tackle those steps during the remainder of this week.


## Testing Instructions

1. Switch to a classic Theme and create a Menu using the `Menus` screen. Add lots of items just for fun.
2. Create another Menu with no items in it. Be sure to save.
3. Switch to block Theme.
4. New Post.
5. Insert a Navigation block.
6. Select the populated classic menu.
7. Immediately see a loading state appear and not sense that the UI is "locking up" (for an understand of what I mean check out the same proceedure on `trunk` or watch the screencaptures below).
8. See the menu converted successfully and inserted into the block.
9. Repeat the same process but this time using the unpopulated classic menu. You should still see the conversion happen but the block will end up being empty because the classic menu was empty. That's expected.
10. Try and test what happens if you go offline just before you click on the classic menu. The UI should handle things and you should see an error notice pop up.
11. Try using screen reader software to run the above steps. You should be able to perceive the progress of the conversion without having to look at the screen. It should be announced to you.

## Screenshots 

### Before

Notice how the dropdown doesn't close and  UI seems to lock up/freeze when I click the classic menu. 

https://user-images.githubusercontent.com/444434/154458859-495a22f5-4138-4dd9-87b1-55ca50706f2e.mp4



### After

Notice the UI lock up is gone. The dropdown disappears and we see the block loading state immediately which resolves as appropriate:

https://user-images.githubusercontent.com/444434/156602370-9478c9fe-c352-4bab-b137-affc5237ebad.mp4

#### Error handling

Notice that the UI communicates the error if the request fails:

https://user-images.githubusercontent.com/444434/156608077-385a689d-ebe2-43fc-a2f3-aa9ed90571d2.mov



## Types of changes





## Checklist:
